mpeg2/src/ts.rs

```diff
@@ -18,85 +20,188 @@ pub struct AdaptationField {
     pub discontinuity_indicator: Option<bool>,
     pub random_access_indicator: Option<bool>,
     pub program_clock_reference_27mhz: Option<u64>,
+    pub private_data_types: Vec<u8>,
+    pub temi_timeline_descriptors: Vec<TEMITimelineDescriptor>,
 }
 
 impl AdaptationField {
     pub fn decode(buf: &[u8]) -> Result<Self, DecodeError> {
+        let mut bs = Bitstream::new(buf);
         let mut af = Self::default();
-        let af_length = buf[0];
-        if af_length as usize > buf.len() - 1 {
-            return Err(DecodeError::new("adaptation field length too long"));
-        } else if buf[0] > 0 {
-            af.discontinuity_indicator = Some(buf[1] & 0x80 != 0);
-            af.random_access_indicator = Some(buf[1] & 0x40 != 0);
-            af.program_clock_reference_27mhz = if af_length >= 7 && buf[1] & 0x10 != 0 {
-                let base = (buf[2] as u64) << 25 | (buf[3] as u64) << 17 | (buf[4] as u64) << 9 | (buf[5] as u64) << 1 | (buf[6] as u64) >> 7;
-                let ext = ((buf[6] as u64) & 1) << 8 | (buf[7] as u64);
+        let af_length = bs.read_u8("adaptation_field_length")?;
+        if af_length > 0 {
+            af.discontinuity_indicator = bs.read_boolean("discontinuity_indicator").ok();
+            af.random_access_indicator = bs.read_boolean("random_access_indicator").ok();
+            bs.skip_bits(1); // elementary_stream_priority_indicator
+
+            let pcr_flag = bs.read_boolean("pcr_flag")?;
+            let opcr_flag = bs.read_boolean("opcr_flag")?;
+            let splicing_point_flag = bs.read_boolean("splicing_point_flag")?;
+            let transport_private_data_flag = bs.read_boolean("transport_private_data_flag")?;
+            let adaptation_extension_flag = bs.read_boolean("adaptation_extension_flag")?;
+
+            af.program_clock_reference_27mhz = if af_length >= 7 && pcr_flag {
+                let base = (bs.read_u32("base high")? as u64) << 1 | bs.read_bit("base low")? as u64;
+                bs.skip_bits(6);
+                let ext = (bs.read_bit("ext high")? as u64) << 8 | bs.read_u8("ext low")? as u64;
                 Some(base * 300 + ext)
             } else {
                 None
             };
+
+            if opcr_flag {
+                // skip OPCR
+                bs.skip_bytes(6);
+            }
+            if splicing_point_flag {
+                // skip splice_countdown
+                bs.skip_bytes(1);
+            }
+            if transport_private_data_flag {
+                let transport_private_data_length = bs.read_u8("transport_private_data_length")?;
+                af.private_data_types = bs.read_n_bytes(transport_private_data_length as usize, "private_data_types")?.to_vec();
```

Comment:
Typo maybe? Was this meant to be "bytes"?

```suggestion
                af.private_data_bytes = bs.read_n_bytes(transport_private_data_length as usize, "private_data_bytes")?.to_vec();
```

Reply:
it's a typo!